**Oksana Harapyn (14711133)
Danylo Redka (14795051)
Sergey Romadin (14988585)
**

In [36]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from sklearn.pipeline import make_pipeline, Pipeline
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import LinearSVC
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import make_classification
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

1.1 Classification on Wine Datase

On the one hand it can be advategeous as multi-class classification aligns better with the discrete nature of wine quality, potentially yielding more interpretable results. Nevertheless, same approach disregards nuances in quality differences, leading to potential loss of information compared to regression.

**2 Portugese Banking Data Analysis**

In [4]:
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")
test_label = pd.read_csv("test_label.csv")

In [8]:
train_noID_Dummies = pd.get_dummies(train.drop(["ID"], axis=1))
test_noID_Dummies = pd.get_dummies(test.drop(["ID"], axis=1))
test_label_noID_Dummies = pd.get_dummies(test_label.drop(["ID"], axis=1))
X_train_unscaled = train_noID_Dummies.drop(["y_no", "y_yes"], axis=1)
X_test_unscaled = test_noID_Dummies
y_train = train_noID_Dummies["y_yes"]
y_test = test_label_noID_Dummies["y_yes"]

In [29]:
pipe_StScaler_LinSVC = make_pipeline(StandardScaler(), LinearSVC())
pipe_MMScaler_LinSVC = make_pipeline(MinMaxScaler(), LinearSVC())

param_grid_LinSVC = {"linearsvc__C": np.logspace(-3, 3, 5)}
grid_StScaler_LinSVC = GridSearchCV(pipe_StScaler_LinSVC, param_grid_LinSVC, cv=10)
grid_MMScaler_LinSVC = GridSearchCV(pipe_MMScaler_LinSVC, param_grid_LinSVC, cv=10)



grid_StScaler_LinSVC.fit(X_train_unscaled, y_train)
grid_MMScaler_LinSVC.fit(X_train_unscaled, y_train)

LinSVC_StScaler_C_best = grid_StScaler_LinSVC.best_params_["linearsvc__C"]
LinSVC_MMScaler_C_best = grid_MMScaler_LinSVC.best_params_["linearsvc__C"]
print("Optimal value for the regularization parameter C using Standard Scaler and LinSVC is: " + str(LinSVC_StScaler_C_best))
print("Optimal value for the regularization parameter C using MinMax Scaler and LinSVC is: " + str(LinSVC_MMScaler_C_best))



LinSVC_StScaler_score = grid_StScaler_LinSVC.score(X_test_unscaled, y_test)
LinSVC_MMScaler_score = grid_MMScaler_LinSVC.score(X_test_unscaled, y_test)
print("The score of the LinearSVC classifier using Standard Scaler is: " + str(LinSVC_StScaler_score))
print("The score of the LinearSVC classifier using MinMax Scaler is: " + str(LinSVC_MMScaler_score))

pipe_unscaled_LinSVC = make_pipeline(LinearSVC())

grid_LinSVC = GridSearchCV(pipe_unscaled_LinSVC, param_grid_LinSVC, cv=10)
grid_LinSVC.fit(X_train_unscaled, y_train)

LinSVC_unscaled_C_best = grid_LinSVC.best_params_["linearsvc__C"]
print("Optimal value for the regularization parameter C using unscaled data and LinSVC is: " + str(LinSVC_unscaled_C_best))

LinSVC_score = grid_LinSVC.score(X_test_unscaled, y_test)
print("The score of the LinearSVC classifier (unscaled data) is: " + str(LinSVC_score))

Optimal value for the regularization parameter C using Standard Scaler and LinSVC is: 0.03162277660168379
Optimal value for the regularization parameter C using MinMax Scaler and LinSVC is: 1.0
The score of the LinearSVC classifier using Standard Scaler is: 0.7762759263574924
The score of the LinearSVC classifier using MinMax Scaler is: 0.7762759263574924
Optimal value for the regularization parameter C using unscaled data and LinSVC is: 1.0
The score of the LinearSVC classifier (unscaled data) is: 0.7760428804474482


**LinearSVC**

The LinearSVC model performs similarly across scaled and unscaled data, with a slight improvement in score when using the Standard Scaler 0.7763 compared to unscaled data 0.7760. The optimal regularization parameter CC varies slightly based on scaling but doesn’t notably affect performance.



In [35]:
pipe_StScaler_LogReg = make_pipeline(StandardScaler(), LogisticRegression(solver='liblinear'))
pipe_MMScaler_LogReg = make_pipeline(MinMaxScaler(), LogisticRegression(solver='liblinear'))

param_grid_LogReg = {"logisticregression__C": np.logspace(-2, 3, 6)}

grid_StScaler_LogReg = GridSearchCV(pipe_StScaler_LogReg, param_grid_LogReg, cv=10, n_jobs=-1)
grid_MMScaler_LogReg = GridSearchCV(pipe_MMScaler_LogReg, param_grid_LogReg, cv=10, n_jobs=-1)

grid_StScaler_LogReg.fit(X_train_unscaled, y_train)
grid_MMScaler_LogReg.fit(X_train_unscaled, y_train)

LogReg_StScaler_C_best = grid_StScaler_LogReg.best_params_["logisticregression__C"]
LogReg_MMScaler_C_best = grid_MMScaler_LogReg.best_params_["logisticregression__C"]
print("Optimal value for the regularization parameter C using Standard Scaler and LogReg is: " + str(LogReg_StScaler_C_best))
print("Optimal value for the regularization parameter C using MinMax Scaler and LogReg is: " + str(LogReg_MMScaler_C_best))

LogReg_StScaler_score = grid_StScaler_LogReg.score(X_test_unscaled, y_test)
LogReg_MMScaler_score = grid_MMScaler_LogReg.score(X_test_unscaled, y_test)
print("The score of the LogisticRegression classifier using Standard Scaler is: " + str(LogReg_StScaler_score))
print("The score of the LogisticRegression classifier using MinMax Scaler is: " + str(LogReg_MMScaler_score))

pipe_unscaled_LogReg = make_pipeline(LogisticRegression(solver='liblinear'))
grid_LogReg = GridSearchCV(pipe_unscaled_LogReg, param_grid_LogReg, cv=10, n_jobs=-1)
grid_LogReg.fit(X_train_unscaled, y_train)

LogReg_unscaled_C_best = grid_LogReg.best_params_["logisticregression__C"]
print("Optimal value for the regularization parameter C using unscaled data and LogReg is: " + str(LogReg_unscaled_C_best))

LogReg_score = grid_LogReg.score(X_test_unscaled, y_test)
print("The score of the LogisticRegression classifier (unscaled data) is: " + str(LogReg_score))


Optimal value for the regularization parameter C using Standard Scaler and LogReg is: 0.01
Optimal value for the regularization parameter C using MinMax Scaler and LogReg is: 100.0
The score of the LogisticRegression classifier using Standard Scaler is: 0.7795385690981124
The score of the LogisticRegression classifier using MinMax Scaler is: 0.7800046609182009
Optimal value for the regularization parameter C using unscaled data and LogReg is: 1000.0
The score of the LogisticRegression classifier (unscaled data) is: 0.7795385690981124


**Logistic Regression**

The Logistic Regression model performs similarly across scaled and unscaled data, with a slight improvement in score when using the MinMax Scaler 0.7800 compared to Standard Scaler 0.7795 and unscaled data 0.7795. The optimal regularization parameter C varies significantly depending on the scaling method.

In [31]:
pipe_StScaler_KNN = make_pipeline(StandardScaler(), KNeighborsClassifier())
pipe_MMScaler_KNN = make_pipeline(MinMaxScaler(), KNeighborsClassifier())

param_grid_KNN = {"kneighborsclassifier__n_neighbors": [3, 5, 7, 9, 11]}

grid_StScaler_KNN = GridSearchCV(pipe_StScaler_KNN, param_grid_KNN, cv=10)
grid_MMScaler_KNN = GridSearchCV(pipe_MMScaler_KNN, param_grid_KNN, cv=10)

grid_StScaler_KNN.fit(X_train_unscaled, y_train)
grid_MMScaler_KNN.fit(X_train_unscaled, y_train)

KNN_StScaler_n_neighbors_best = grid_StScaler_KNN.best_params_["kneighborsclassifier__n_neighbors"]
KNN_MMScaler_n_neighbors_best = grid_MMScaler_KNN.best_params_["kneighborsclassifier__n_neighbors"]
print("Optimal value for n_neighbors using Standard Scaler and KNN is: " + str(KNN_StScaler_n_neighbors_best))
print("Optimal value for n_neighbors using MinMax Scaler and KNN is: " + str(KNN_MMScaler_n_neighbors_best))

KNN_StScaler_score = grid_StScaler_KNN.score(X_test_unscaled, y_test)
KNN_MMScaler_score = grid_MMScaler_KNN.score(X_test_unscaled, y_test)
print("The score of the KNeighborsClassifier classifier using Standard Scaler is: " + str(KNN_StScaler_score))
print("The score of the KNeighborsClassifier classifier using MinMax Scaler is: " + str(KNN_MMScaler_score))

pipe_unscaled_KNN = make_pipeline(KNeighborsClassifier())
grid_KNN = GridSearchCV(pipe_unscaled_KNN, param_grid_KNN, cv=10)
grid_KNN.fit(X_train_unscaled, y_train)

KNN_unscaled_n_neighbors_best = grid_KNN.best_params_["kneighborsclassifier__n_neighbors"]
print("Optimal value for n_neighbors using unscaled data and KNN is: " + str(KNN_unscaled_n_neighbors_best))

KNN_score = grid_KNN.score(X_test_unscaled, y_test)
print("The score of the KNeighborsClassifier classifier (unscaled data) is: " + str(KNN_score))

Optimal value for n_neighbors using Standard Scaler and KNN is: 11
Optimal value for n_neighbors using MinMax Scaler and KNN is: 11
The score of the KNeighborsClassifier classifier using Standard Scaler is: 0.7718480540666511
The score of the KNeighborsClassifier classifier using MinMax Scaler is: 0.7653227685854114
Optimal value for n_neighbors using unscaled data and KNN is: 11
The score of the KNeighborsClassifier classifier (unscaled data) is: 0.7152178979258914


**KNeighborsClassifier**

The KNeighborsClassifier model performs best with # of neighbous is 11 for both Standard and MinMax scaling. The Standard Scaler yields the highest score 0.7718, followed by MinMax Scaler 0.7653, and unscaled data 0.715. Scaling improves performance significantly compared to unscaled data.

In [38]:


dt_model = DecisionTreeClassifier()

dt_model.fit(X_train_unscaled, y_train)
dt_accuracy = dt_model.score(X_test_unscaled, y_test)

print("Decision Tree Accuracy: ", dt_accuracy)

y_pred_dt = dt_model.predict(X_test_unscaled)

print("Classification Report for Decision Tree:\n", classification_report(y_test, y_pred_dt))
print("Confusion Matrix for Decision Tree:\n", confusion_matrix(y_test, y_pred_dt))

param_grid_dt = {'max_depth': [3, 5, 10, None],'max_features': ['sqrt', 'log2', None]}
grid_dt = GridSearchCV(dt_model, param_grid_dt, cv=10)

grid_dt.fit(X_train_unscaled, y_train)
print("Best Hyperparameters for Decision Tree: ", grid_dt.best_params_)
print("Best Decision Tree Model Accuracy: ", grid_dt.best_score_)

Decision Tree Accuracy:  0.6790957818690282
Classification Report for Decision Tree:
               precision    recall  f1-score   support

       False       0.78      0.75      0.76      2969
        True       0.48      0.52      0.50      1322

    accuracy                           0.68      4291
   macro avg       0.63      0.64      0.63      4291
weighted avg       0.69      0.68      0.68      4291

Confusion Matrix for Decision Tree:
 [[2225  744]
 [ 633  689]]
Best Hyperparameters for Decision Tree:  {'max_depth': 10, 'max_features': None}
Best Decision Tree Model Accuracy:  0.7675990675990676


**Decision Tree**

The Decision Tree model shows a relatively high accuracy of 0.68, but its performance on the minority class "True" is weaker with a recall of 0.52. The grid search suggests that a maximum depth of 10 and no restriction on features yield the best performance, improving the model's accuracy to 0.77. This indicates that while the model is prone to underperforming on imbalanced data, tuning can help optimize results.